[RFC] Send LOGICAL_ERRORs to sentry

azat opened the pull request:
**Thoughts on this one?**

This is like an assert() in the code, so it is useful to know about them
as well.

This is not enabled by default, to turn it ON, set `send_crash_reports.send_logical_errors=true` in config

### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

azat committed:
Send LOGICAL_ERRORs to sentry

This is like an assert() in the code, so it is useful to know about them
as well.

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Rewrite SentryWriter to a singleton

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

robot-ch-test-poll3 commented:
<!-- automatic status comment for PR #61997 from azat/ClickHouse:sentry-logical-error -->
*This is an automated comment for commit 2652ab36a26bec5c560ffb88470af0d62745608e with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/61997/2652ab36a26bec5c560ffb88470af0d62745608e/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

nikitamikhaylov commented:
Ok, but let it be configurable at least https://clickhouse.com/docs/en/operations/server-configuration-parameters/settings#send_crash_reports

azat committed:
Avoid SentryWriter deps for Exception

    $ ninja unit_tests_dbms
    FAILED: src/unit_tests_dbms
    : && /usr/bin/clang++ --target=x86_64-linux-gnu --sysroot=/src/ch/clickhouse/cmake/linux/../../contrib/sysroot/linux-x86_64/x86_64-linux-gnu/libc --gcc-toolchain=/src/ch/clickhouse/cmake/linux/../../contrib/sysroot/linux-x86_64 -fdiagnostics-color=always -Xclang -fuse-ctor-homing -Wno-enum-constexpr-conversion -fsized-deallocation  -gdwarf-aranges -pipe -mssse3 -msse4.1 -msse4.2 -mpclmul -mpopcnt -fasynchronous-unwind-tables -falign-functions=32 -mbranches-within-32B-boundaries -ffp-contract=off  -fno-omit-frame-pointer -mno-omit-leaf-frame-pointer -fdiagnostics-absolute-paths -fstrict-vtable-pointers -Wall -Wextra -Wframe-larger-than=65536 -Weverything -Wpedantic -Wno-zero-length-array -Wno-c++98-compat-pedantic -Wno-c++98-compat -Wno-c++20-compat -Wno-sign-conversion -Wno-implicit-int-conversion -Wno-implicit-int-float-conversion -Wno-ctad-maybe-unsupported -Wno-disabled-macro-expansion -Wno-documentation-unknown-command -Wno-double-promotion -Wno-exit-time-destructors -Wno-float-equal -Wno-global-constructors -Wno-missing-prototypes -Wno-missing-variable-declarations -Wno-padded -Wno-switch-enum -Wno-undefined-func-template -Wno-unused-template -Wno-vla -Wno-weak-template-vtables -Wno-weak-vtables -Wno-thread-safety-negative -Wno-enum-constexpr-conversion -Wno-unsafe-buffer-usage -Wno-switch-default -O2 -g -DNDEBUG -O3 -g  -fno-omit-frame-pointer -mno-omit-leaf-frame-pointer -fno-pie --gcc-toolchain=/src/ch/clickhouse/cmake/linux/../../contrib/sysroot/linux-x86_64 --ld-path=/usr/bin/ld.lld -Wl,--no-export-dynamic -Wl,--gc-sections -Wl,--gdb-index -Wl,--build-id=sha1 -no-pie -Wl,-no-pie    -Xlinker --no-undefined @CMakeFiles/unit_tests_dbms.rsp -o src/unit_tests_dbms && :
    ld.lld: error: undefined symbol: SentryWriter::getInstance()
    >>> referenced by Exception.cpp:63 (/src/ch/clickhouse/src/Common/Exception.cpp:63)
    >>>               Exception.cpp.o:(DB::handle_error_code(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, int, bool, std::__1::vector<void*, std::__1::allocator<void*>> const&)) in archive src/libclickhouse_common_io.a
    >>> referenced by Exception.cpp:63 (/src/ch/clickhouse/src/Common/Exception.cpp:63)
    >>>               Exception.cpp.o:(DB::Exception::Exception(DB::Exception::MessageMasked const&, int, bool)) in archive src/libclickhouse_common_io.a
    >>> referenced by Exception.cpp:63 (/src/ch/clickhouse/src/Common/Exception.cpp:63)
    >>>               Exception.cpp.o:(DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool)) in archive src/libclickhouse_common_io.a

    ld.lld: error: undefined symbol: SentryWriter::onFault(int, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::array<void*, 45ul> const&, unsigned long, unsigned long)
    >>> referenced by Exception.cpp:68 (/src/ch/clickhouse/src/Common/Exception.cpp:68)
    >>>               Exception.cpp.o:(DB::handle_error_code(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, int, bool, std::__1::vector<void*, std::__1::allocator<void*>> const&)) in archive src/libclickhouse_common_io.a
    >>> referenced by Exception.cpp:68 (/src/ch/clickhouse/src/Common/Exception.cpp:68)
    >>>               Exception.cpp.o:(DB::Exception::Exception(DB::Exception::MessageMasked const&, int, bool)) in archive src/libclickhouse_common_io.a
    >>> referenced by Exception.cpp:68 (/src/ch/clickhouse/src/Common/Exception.cpp:68)
    >>>               Exception.cpp.o:(DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool)) in archive src/libclickhouse_common_io.a
    clang++: error: linker command failed with exit code 1 (use -v to see invocation)
    ninja: build stopped: subcommand failed.

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Make sending LOGICAL_ERRORs to sentry configurable

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat commented:
Make sense - done. Configurable with `send_crash_reports.send_logical_errors`

azat commented:
>Stateless tests (asan) [1/4] — Tests are not finished, fail: 2, passed: 564, skipped: 8
>Stateless tests (ubsan) [1/2] — Tests are not finished, fail: 4, passed: 1145, skipped: 14

- https://github.com/ClickHouse/ClickHouse/issues/62062

azat commented:
Can someone take a look please?

antaljanosbenjamin commented:
I can take a look at this today.

antaljanosbenjamin commented on src/Daemon/SentryWriter.cpp:
```diff
@@ -222,8 +237,11 @@ void SentryWriter::onFault(int sig, const std::string & error_message, const Sta
-void SentryWriter::initialize(Poco::Util::LayeredConfiguration &) {}
-void SentryWriter::shutdown() {}
-void SentryWriter::onFault(int, const std::string &, const StackTrace &) {}
+void SentryWriter::initializeInstance(Poco::Util::LayeredConfiguration &) {}
+SentryWriter * SentryWriter::getInstance() { return nullptr; }
+
+SentryWriter::SentryWriter(Poco::Util::LayeredConfiguration &) {}
+SentryWriter::~SentryWriter() = default;
+void SentryWriter::onFault(int, const std::string &, const FramePointers &, size_t, size_t) {}
```
```suggestion
```

azat replied:
This placeholders required when sentry is disabled.

azat replied:
Here is an example - https://s3.amazonaws.com/clickhouse-test-reports/61997/8b39cb2c6d08a2e86867e068bef0a8689362634a/fast_test/run.log

```
2024-04-11 17:21:08 /ClickHouse/src/Daemon/SentryWriter.cpp:260:20: error: out-of-line definition of 'onFault' does not match any declaration in 'SentryWriter'
2024-04-11 17:21:08   260 | void SentryWriter::onFault(int, const std::string &, const FramePointers &, size_t, size_t) {}
2024-04-11 17:21:08       |                    ^~~~~~~

```

antaljanosbenjamin replied:
Oh, I totally missed this, you are right.

antaljanosbenjamin commented on src/Daemon/SentryWriter.cpp:
```diff
@@ -133,35 +141,43 @@ void SentryWriter::initialize(Poco::Util::LayeredConfiguration & config)
         sentry_value_t event = sentry_value_new_message_event(SENTRY_LEVEL_FATAL, "fault", error_message.c_str());
-        sentry_set_tag("signal", strsignal(sig)); // NOLINT(concurrency-mt-unsafe) // not thread-safe but ok in this context
-        sentry_set_extra("signal_number", sentry_value_new_int32(sig));
+        if (sig_or_error > 0)
+        {
+            int sig = sig_or_error;
+            sentry_set_tag("signal", strsignal(sig)); // NOLINT(concurrency-mt-unsafe) // not thread-safe but ok in this context
+            sentry_set_extra("signal_number", sentry_value_new_int32(sig));
+        }
+        else
+        {
+            /// Can be only LOGICAL_ERROR, but just in case.
+            int code = -sig_or_error;
+            sentry_set_tag("exception", DB::ErrorCodes::getName(code).data());
+            sentry_set_extra("exception_code", sentry_value_new_int32(code));
+        }
```
I think it is better to separate not just by the sign of the things. There are two possibilities in my opinion:
1. keep thet `onFault` for signals and make another function for exceptions
2. introduce an enum/boolean flag to `onFault` that indicates whether it is signal or error

azat replied:
Done.

antaljanosbenjamin commented on src/Daemon/SentryWriter.cpp:
```diff
@@ -133,35 +141,43 @@ void SentryWriter::initialize(Poco::Util::LayeredConfiguration & config)
 }
 
-void SentryWriter::onFault(int sig, const std::string & error_message, const StackTrace & stack_trace)
+void SentryWriter::onFault(int sig_or_error, const std::string & error_message, const FramePointers & frame_pointers, size_t offset, size_t size)
```
I couldn't find definite answer in sentry docs about thread-safety. [They](https://docs.sentry.io/platforms/native/performance/instrumentation/custom-instrumentation/#thread-safety) mentioned that if something is not thread safe, then it is mentioned in the function's docs. `sentry_value_new_message_event` doesn't mention that, so my assumption is it is thread-safe. If I am right, then it is fine, if not, then we should make sure two logical errors happening at the same time won't cause any issues.

Previously we only reported crashes, so I am not sure if anybody thought about thread-safety, therefore I think we should make sure it is thread safe.

azat replied:
Yes, it looks safe, global variables guarded with mutex (thought 100% it is hard to tell of course).

antaljanosbenjamin commented on src/Daemon/SentryWriter.h:
```diff
@@ -14,14 +14,32 @@ class StackTrace;
 /// It is possible to send those reports to your own sentry account or account of consulting company you hired
 /// by overriding "send_crash_reports.endpoint" setting. "send_crash_reports.debug" setting will allow to do that for
-namespace SentryWriter
+class SentryWriter
```
As I am trying to be cautious about what data I share with companies, I think it is a must to mention that this class from now on will also send some exceptions too. Could you please extend its documentation briefly?

azat replied:
Documentation for the class you mean?

Documentation itself is already updated:

```
`send_logical_errors` – `LOGICAL_ERROR` is like an `assert`, it is a bug in ClickHouse. This boolean flag enables sending this exceptions to sentry (default: `false`).
```

azat replied:
Updated brief comment for class as well.

antaljanosbenjamin replied:
> Documentation for the class you mean?

Yes, I meant that, thanks for updating it.

antaljanosbenjamin reviewed (changes requested):
The PR in general is good in my opinion. There are a few things to change, but nothing serious.

The only question I have remaining is have you tried it out at least once? If not, then I can make an attempt to see at least one `LOGICAL_ERROR` sent to sentry.

azat committed:
More cleaner SentryWriter API

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat commented:
@antaljanosbenjamin does it waits something else?

antaljanosbenjamin commented:
Sorry, I forgot to merge it. Merging now.
